driver-async/src/test/functional/com/mongodb/async/client/RetryableWritesProseTest.java

```diff
@@ -0,0 +1,245 @@
+public class RetryableWritesProseTest extends DatabaseTestCase {
+    private MongoClient clientUnderTest;
+    private MongoClient failPointClient;
+    private MongoClient stepDownClient;
+    private final String databaseName;
+    private final String collectionName;
+    private final String description;
+
+    public RetryableWritesProseTest() {
+        this.description = "Replica set failover test for retryable writes";
+        this.databaseName = getDefaultDatabaseName();
+        this.collectionName = "test";
+    }
+
+    @BeforeClass
+    public static void beforeClass() {
+    }
+
+    @AfterClass
+    public static void afterClass() {
+    }
+
+    @Before
+    @Override
+    public void setUp() {
+        assumeTrue(canRunTests());
+
+        setUpClientUnderTest();
+        setUpFailPointClient();
+        setUpStepDownClient();
+    }
+
+    @After
+    public void cleanUp() {
+        if (failPointClient != null) {
+            MongoDatabase failPointAdminDB = failPointClient.getDatabase("admin");
+            FutureResultCallback<Document> futureResultCallback = new FutureResultCallback<Document>();
+            failPointAdminDB.runCommand(
+                    new Document("configureFailPoint", "stepdownHangBeforePerformingPostMemberStateUpdateActions")
+                            .append("mode", "off"), futureResultCallback);
+        }
+
+        if (clientUnderTest != null) {
+            clientUnderTest.close();
+        }
+        if (stepDownClient != null) {
+            stepDownClient.close();
+        }
+    }
+
+    @Test
+    public void shouldPassAllOutcomes() {
+        // Using the client under test, insert a document and observe a successful write result.
+        // This will ensure that initial discovery takes place.
+        MongoCollection<Document> collection = insertInitialDocument();
+
+        // Using the fail point client, activate the fail point by setting mode to "alwaysOn".
+        activateFailPoint();
+
+        // Using the step down client, step down the primary by executing the command { replSetStepDown: 60, force: true}.
+        // This operation will hang so long as the fail point is activated. When the fail point is later deactivated,
+        // the step down will complete and the primary's client connections will be dropped. At that point, any ensuing
+        // network error should be ignored.
+        MongoDatabase stepDownDB = stepDownPrimary();
+
+        // Using the client under test, insert a document and observe a successful write result. The test MUST assert
+        // that the insert command fails once against the stepped down node and is successfully retried on the newly
+        // elected primary (after SDAM discovers the topology change). The test MAY use APM or another means to observe
+        // both attempts.
+        insertDocumentShouldFail(collection);
```

Comment:
This should succeed - because part of what the test is aiming to confirm is retrying after there has been a stepdown.

If you add the `TestCommandListener`  to the `ClientUnderTest` then you should be able to assert there haven't been any errors before trying to insert and then the insert should succeed but you should be able to observe an error in the events from the listener.

Reply:
I have added the `TestCommandListener` to the client under test and changed the `insertDocumentShouldFail()` method to just `insertDocument()`. The test passes now, but I am not able to observe a `notMaster` error. However, I do see the following in the log:
```
11:02:26.164 [main] INFO  org.mongodb.driver.cluster - No server chosen by WritableServerSelector from cluster description 
```
Would this be expected after step down?